File: apps/server/src/routes/account.ts
import { Router, Request, Response, NextFunction } from "express";
import { deviceAuthMiddleware } from "../middleware/deviceAuth";
import * as accountService from "../services/accountService";

const router = Router();

const asyncHandler =
  (fn: (req: Request, res: Response, next: NextFunction) => Promise<void>) =>
  (req: Request, res: Response, next: NextFunction) =>
  fn(req, res, next).catch(next);

router.use(deviceAuthMiddleware);

/**
 * GET /wallets/networks/batch — Batch get account networks for multiple wallets
 * Returns deduplicated network list per wallet, for lightweight UI display
 * Uses wallets_addresses table (replaces old accounts table)
 */
router.get(
  "/wallets/networks/batch",
  asyncHandler(async (req: Request, res: Response) => {
    const walletIds = (req.query.walletIds as string || "").split(",").filter(Boolean);
    if (walletIds.length === 0) {
      res.json({ wallets: [] });
      return;
    }

    const result = await accountService.getWalletsNetworksBatch(walletIds);
    res.json({ wallets: result });
  })
);

/**
 * GET /chains/available — 获取支持创建账户的链列表
 */
router.get("/chains/available", asyncHandler(async (_req: Request, res: Response) => {
  const chains = await accountService.getAvailableChains();
  res.json({ chains });
}));

export default router;

File: apps/server/src/routes/account.rs
//! 账户路由 — /api/v1/accounts
//! 迁移自 IMWallet routes/account.ts (2 个接口)

use crate::errors::AppError;
use crate::middleware::AppState;
use crate::services::account_service;
use axum::{
    extract::{Query, State},
    routing::get,
    Json, Router,
};
use serde::{Deserialize, Serialize};

pub fn router() -> Router<AppState> {
    Router::new()
        .route("/accounts/chains/available", get(get_available_chains))
        .route("/accounts/wallets/networks/batch", get(get_wallet_networks))
}

#[derive(Debug, Serialize)]
struct ChainsResponse {
    chains: Vec<account_service::AvailableChain>,
}

/// GET /accounts/chains/available — 获取支持创建账户的链列表
async fn get_available_chains(
    State(state): State<AppState>,
) -> Result<Json<ChainsResponse>, AppError> {
    let chains = account_service::get_available_chains_cached(state.db.clone()).await?;
    Ok(Json(ChainsResponse { chains }))
}

#[derive(Debug, Deserialize)]
pub struct WalletNetworksParams {
    #[serde(rename = "walletIds")]
    pub wallet_ids: String,
}

#[derive(Debug, Serialize)]
struct WalletsNetworkResponse {
    wallets: Vec<account_service::WalletNetworkInfo>,
}

async fn get_wallet_networks(
    State(state): State<AppState>,
    Query(params): Query<WalletNetworksParams>,
) -> Result<Json<WalletsNetworkResponse>, AppError> {
    let ids: Vec<String> = params
        .wallet_ids
        .split(',')
        .map(|s| s.trim().to_string())
        .collect();
    let wallets = account_service::get_wallet_networks(state.db.clone(), &ids).await?;
    Ok(Json(WalletsNetworkResponse { wallets }))
}
